docs/adr/012_decouple_from_edge.md

```diff
@@ -0,0 +1,147 @@
+
+### In-scope
+
+- Describe how to differentiate "edge" LNSs deployments from "cloud" ones in mixed scenarios
+- Provide an alternative for deploying and managing configuration
+- Provide additional information on abstracting and replacing ModuleClient related dependencies, such as:
+  - Direct methods (currently used for Clearing Cache, forcefully closing DeviceClient connections, handling class-c c2d messages)
+  - Twin updates (currently used for updating few configuration items, without restarting the entire LNS)
+- Provide information on a "Function" endpoint for clearing cache of all "cloud" LNSs
+- Describe a sample deployment scenario being enabled by decoupling
+
+### Out of scope
+
+- Key Vault integration for secrets management  
+It would be great to make use of Azure Key Vault to securely store components like Azure Function authentication code or Server Certificate for Secure BasicsStation-LNS communications, but this document will not focus on this aspect.
+
+## Changes required
+
+Before diving in the various required changes, let's understand the "IoT Edge" bindings that are there in the code.
+
+The main difference comes to "ModuleConnectionHost", which is responsible for:
+
+- Instantiating a [ModuleClient](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.devices.client.moduleclient?view=azure-dotnet) object to be used for interacting with module twins and receiving direct method invocations
+- Setting a callback on "desired property updates" for dynamically configuring some configuration variables
+- Setting a callback on "direct method invocation" for handling:
+  - Clearing LNS device registry cache
+  - Closing DeviceClient connections for avoiding "ping-pong" as described in [ADR 010](./010_lns_affinity.md)
+  - Sending Cloud to Device messages for Class-C devices
+
+Except for ModuleConnectionHost, the other difference is concerning the ability to "proxy" leaf-device messages through Edge Hub. By default the "ENABLE_GATEWAY" environment variable is set to true, we need to make sure that **ENABLE_GATEWAY can't be set to true when not running as Edge module**
+
+### Differentiate Edge from Cloud deployments
+
+As of v2.1.0, the LNS code is already able to understand whether it is running as IoT Edge Module or not. The availability of the "IOTEDGE_APIVERSION" environment variable is checked. This variable is injected by the "Edge Agent" component at the moment of module creation and startup, therefore if the variable is not there we can safely assume that the LNS is not running in a Edge environment.
+
+When a leaf device message is sent and processed, the device twin gets updated with the "GatewayID", currently being set to the "IOTEDGE_DEVICEID" environment variable.
```

Comment:
Can we rely on that? What if they remove this or change it? Should we have an explicit variable we provision to indicate the deployment model?

Reply:
To be honest, IOTEDGE_APIVERSION has been there since the very beginning of IoT Edge, I am not expecting them to deprecate this soon. At the same time, it's not something that is meant for "public" usage, therefore we might consider adding an additional variable to be used for identifying a "CLOUD_DEPLOYMENT"